File: command/command/history/History.cs
using command.commands;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace command.history
{
	public class History : IHistory
	{
		private readonly IList<AbstractCommand> _commands = new List<AbstractCommand>();
		private readonly uint _maxCommandsCount = uint.MaxValue;
		private  int _executedCommandIndex = -1;

		public History(uint maxCommandsCount = uint.MaxValue)
		{
			_maxCommandsCount = maxCommandsCount;
		}

		public bool CanUndo
		{
			get
			{
				return _commands.Count != 0 && _executedCommandIndex > -1;
			}
		}

		public bool CanRedo
		{
			get
			{
				return _commands.Count != 0 && _executedCommandIndex < _commands.Count - 1;
			}
		}

		/// <summary>
		/// Adds the command.
		/// Command must be already executed
		/// </summary>
		/// <param name="command">The command.</param>
		public void AddCommand(AbstractCommand command)
		{
			if (command == null)
			{
				throw new ArgumentNullException(nameof(command));
			}
			if (!command.IsExecuted)
			{
				throw new ArgumentException(nameof(command));
			}

			RemoveUnexecutedCommands();
			RemoveOldCommands();

			_commands.Add(command);
			_executedCommandIndex = _commands.Count - 1;
		}

		public void Dispose()
		{
			foreach(var command in _commands)
			{
				command.Dispose();
			}
			_commands.Clear();
		}

		public void Redo()
		{
			if (!CanRedo)
			{
				throw new InvalidOperationException("nothing to redo");
			}

			_commands[++_executedCommandIndex].Execute();
		}

		public void Undo()
		{
			if (!CanUndo)
			{
				throw new InvalidOperationException("nothing to undo");
			}

			_commands[_executedCommandIndex--].Unexecute();
		}

		private void RemoveOldCommands()
		{
			if (_commands.Count == _maxCommandsCount)
			{
				_commands[0].Dispose();
				_commands.RemoveAt(0);
			}
		}

		private void RemoveUnexecutedCommands()
		{
			if (CanRedo)
			{
				for (var i = _commands.Count - 1; i > _executedCommandIndex; --i)
				{
					_commands[i].Dispose();
					_commands.RemoveAt(i);
				}
			}
		}
	}
}

File: command_java/Document/src/hsitory/History.java
package hsitory;

import java.util.ArrayList;
import java.util.List;

public class History implements IHistory, AutoCloseable {

    public History(int historyLength) {
        _historyLength = historyLength;
    }

    @Override
    public void AddCommand(ICommand command) throws Exception {
        if (!command.IsExecuted())
        {
            throw new IllegalArgumentException("Command must be already executed");
        }

        RemoveUnExecutedCommands();
        RemoveOldCommands();
        AddNewCommand(command);
    }

    @Override
    public void Undo() {
        if (!CanUndo())
        {
            throw new IllegalStateException("Nothing to undo");
        }

        _commands.get(_executedCommandIndex--).UnExecute();
    }

    @Override
    public void Redo() {
        if (!CanRedo())
        {
            throw new IllegalStateException("Nothing to redo");
        }

        _commands.get(++_executedCommandIndex).Execute();
    }

    @Override
    public void close() throws Exception {
        for(ICommand command: _commands)
        {
            command.Destroy();
        }
    }

    private void AddNewCommand(ICommand command) {
        _commands.add(command);
        _executedCommandIndex = _commands.size() - 1;
    }

    private void RemoveOldCommands() throws Exception {
        if (_commands.size() == _historyLength)
        {
            _commands.get(0).Destroy();
            _commands.remove(0);
        }
    }
    private void RemoveUnExecutedCommands() throws Exception {
        if (CanRedo())
        {
            for (int i = _commands.size() - 1; i > _executedCommandIndex; --i)
            {
                _commands.get(i).Destroy();
                _commands.remove(i);
            }
        }
    }

    private boolean CanUndo() {
        return _commands.size() != 0 && _executedCommandIndex > -1;
    }

    private boolean CanRedo() {
        return _commands.size() != 0 && _executedCommandIndex < _commands.size() - 1;
    }

    private  int _executedCommandIndex = -1;
    private int _historyLength = 10;
    private final List<ICommand> _commands = new ArrayList<>();
}
